src/pytest_workflow/util.py

```diff
@@ -211,6 +212,38 @@ def file_md5sum(filepath: Path, block_size=64 * 1024) -> str:
     return hasher.hexdigest()
 
 
+def file_diff(first_filepath: Path, second_filepath: Path,
+              encoding: Optional[str] = None,
+              max_lines: int = 20) -> Optional[str]:
+    """
+    Generates a diff between two files.
+    :param first_filepath: a pathlib.Path to the first file
+    :param second_filepath: a pathlib.Path to the second file
+    :param encoding: a pathlib.Path to the second file.
+    :param max_lines: maximum number of diff lines to include
+    :return: a unified diff of the two files, or None if they are identical
+    """
+    if encoding is None:
+        encoding = sys.getdefaultencoding()
+
+    with first_filepath.open('r', encoding=encoding) as first_file_handler:
+        first_lines = first_file_handler.read().splitlines()
+    with second_filepath.open('r', encoding=encoding) as second_file_handler:
+        second_lines = second_file_handler.read().splitlines()
+
+    diff_lines = list(difflib.unified_diff(
+        first_lines, second_lines,
+        fromfile=str(first_filepath),
+        tofile=str(second_filepath),
+        lineterm=''
+    ))
+    if len(diff_lines) > max_lines:
+        diff_lines = diff_lines[:max_lines]
+        diff_lines.append('[...]')
+
+    return '\n'.join(diff_lines) if diff_lines else None
```

Comment:
This is a slight anti-pattern, first reading everything into memory, then only selecting the first `max_lines`.

The code can use less memory with the following:
```suggestion
    diff_lines = difflib.unified_diff(
        first_lines, second_lines,
        fromfile=str(first_filepath),
        tofile=str(second_filepath),
        lineterm=''
    )
    shortened_diff_lines = [line for _, line in zip(range(max_lines), diff_lines]
    try:  # Check if the diff has more lines
        next(diff_lines)
        shortened_diff_lines.append('[...]')
    except StopIteration:
        pass
     return '\n'.join(diff_lines)
    ```
Also it makes sense to return a str here, not a None. See my other comment.

Reply:
addressed in 476eb831688aa7a6ecd52e2e39299f0eb99de243.

fun fact, I tried passing file line generators into `difflib.unified_diff` to avoid reading each file into memory but there are difflib [type checks that index into the inputs](https://github.com/python/cpython/blob/main/Lib/difflib.py#L1263-L1271)